### PR TITLE
Tighten up our compression framework.

### PR DESCRIPTION
**Pull request by jpountz**
We have a compression framework that we use internally, mainly to compress some
xcontent bytes. However it is quite lenient: for instance it relies on the
assumption that detection of the compression format can only be called on either
some compressed xcontent bytes or some raw xcontent bytes, but nothing checks
this. By the way, we are misusing it in BinaryFieldMapper so that if someone
indexes a binary field which happens to have the same header as a LZF stream,
then at read time, we will try to decompress it.

It also simplifies the API by removing block compression (only streaming) and
some code duplication caused by some methods accepting a byte[] and other
methods a BytesReference.

### CODE REVIEW

**Comment by jpountz** [comment]
I'm reading the description again and I've probably not been clear enough about what changed:
 - If you call CompressorFactory.compressor(ChannelBuffer) and the compression format is not detected, you will get an exception. This was already working this way at the call site: MessageChannelHandler.messageReceived, now it's done directly in the framework.
 - If you call CompressorFactory.compressor(BytesReference) and the bytes are not either some xcontent bytes or some compressed xcontent bytes, then you will get an exception. There is no way in general to detect if some bytes are compressed because nothing prevents an arbitrary bytes reference to have the same header as a LZF block so we enforce that this method be only used on xcontent bytes, which is how we are using this method (to the notable exception of BinaryFieldMapper, which is buggy as described in the description).

**Comment by jpountz** [comment]
I opened #11280 to fix the BinaryFieldMapper issue.

**Comment by jpountz** [comment]
I pushed a new commit to also use DEFLATE (with a compression level of 3) instead of LZF. The reason is that even if LZF might be faster, we have had some issues recently like https://github.com/elastic/elasticsearch/pull/7210 or https://github.com/elastic/elasticsearch/pull/7468 and expect that using something that is more widely used like DEFLATE will protect us better from corruptions in the future.

File: src/main/java/org/elasticsearch/common/compress/NotXContentException.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.compress;
+
+/** Exception indicating that we were expecting something compressed, which
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/common/compress/NotXContentException.java]: copy-pasted from NotCompressedException?

File: src/main/java/org/elasticsearch/common/compress/deflate/DeflateCompressor.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            if (read == -1) {
+                break;
+            }
+            len += read;
+        }
+        if (len != HEADER.length || Arrays.equals(headerBytes, HEADER) == false) {
+            throw new IllegalArgumentException("Input stream is not compressed with DEFLATE!");
+        }
+
+        final boolean nowrap = true;
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/common/compress/deflate/DeflateCompressor.java]: This is in a local just for readability?

File: src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java [hunk 3]
```diff
@@ -457,9 +416,11 @@ public static void writeDirect(BytesReference source, XContentBuilder rawBuilder
     public static void writeRawField(String field, BytesReference source, XContentBuilder builder, ToXContent.Params params) throws IOException {
         Compressor compressor = CompressorFactory.compressor(source);
         if (compressor != null) {
-            CompressedStreamInput compressedStreamInput = compressor.streamInput(source.streamInput());
+            InputStream compressedStreamInput = compressor.streamInput(source.streamInput());
             XContentType contentType = XContentFactory.xContentType(compressedStreamInput);
-            compressedStreamInput.resetToBufferStart();
+            if (compressedStreamInput.markSupported() == false) {
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java]: Shouldn't this be before the call to XContentFactory.xContentType?

File: src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java [hunk 4]
```diff
@@ -144,10 +147,16 @@ public BytesReference value(Object value) {
         }
         try {
             if (indexCreatedBefore2x) {
-                return CompressorFactory.uncompressIfNeeded(bytes);
-            } else {
-                return bytes;
+                try {
+                    return CompressorFactory.uncompressIfNeeded(bytes);
+                } catch (NotXContentException e) {
+                    // This is a BUG! We try to decompress by detecting a header in
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java]: This is scary to see a comment saying "This is a BUG!" and then doing nothing with the exception...

**Comment by rjernst** [comment]
LGTM, I left a couple minor comments.

File: src/main/java/org/elasticsearch/common/compress/deflate/DeflateCompressor.java (hunk 2 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/compress/deflate/DeflateCompressor.java]: exactly

File: src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java (hunk 4 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java]: Right, it's unfortunate we need it for backward compatibility. Would you like maybe a warning in that case?

Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/core/BinaryFieldMapper.java]: Yes maybe add a warning, and add to the comment that the buggy behavior is kept for backwards compatibility?

**Comment by jpountz** [comment]
@rjernst Thanks for the review, I pushed a new commit.

**Comment by rjernst** [comment]
LGTM

**Pull request merged**